### PR TITLE
Refine InstructorFeedbackResults slice #3433

### LINKED ISSUE #3433: Refine InstructorFeedbackResults slice
Java main side:
InstructorFeedbackResultsDownloadAction
InstructorFeedbackResultsPageAction
InstructorFeedbackResultsPageData
InstructorFeedbackResultsResponseRatePageAction
InstructorFeedbackResultsResponseRatePageData
FeedbackSessionResultsBundle
FeedbackSessionsResponseStatus

Java testing side:
InstructorFeedbackResultsDownloadActionTest
InstructorFeedbackResultsPageActionTest
InstructorFeedbackResultsPageUiTest
InstructorFeedbackResultsPage

JS:
instructorFeedbackResults.js
instructorFeedbackResultsAjaxByGQR.js
instructorFeedbackResultsAjaxByGRQ.js
instructorFeedbackResultsAjaxByQuestion.js
instructorFeedbackResultsAjaxByRGQ.js
instructorFeedbackResultsAjaxByRQG.js
instructorFeedbackResultsAjaxResponseRate.js
InstructorFeedbackResultsPageUiTest.json

JSP and HTML to be done separately because of their ridiculously huge size.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #3433 
So far only the Java main files are done.

### CODE REVIEW

**Comment by wkurniawan07** [comment]
Rebased to revert all the JSP file changes. They are still stashed in my local repo. This slice will be focused on Java coding style and tests/coverage, and JS since there aren't many changes there.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java [hunk 1]
```diff
@@ -106,70 +97,68 @@ protected ActionResult execute() throws EntityDoesNotExistException {
...
+                                                                               queryRange, data.sortType);
         } else if (data.sortType.equals("question")) {
-            if(questionNumStr == null){
-                data.bundle = logic.getFeedbackSessionResultsForInstructorInSection(
-                                feedbackSessionName, courseId,
-                                data.instructor.email, data.selectedSection);
+            if (questionNumStr == null) {
+                data.bundle = logic.getFeedbackSessionResultsForInstructorInSection(feedbackSessionName, courseId,
+                                                                                    data.instructor.email,
+                                                                                    data.selectedSection);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: This is OK too (although it breaks the line wrapping rules, the readability remains similar while we save a line).
```
data.bundle = logic.getFeedbackSessionResultsForInstructorInSection(feedbackSessionName, courseId,
                                                          data.instructor.email,  data.selectedSection);
```

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 2]
```diff
@@ -24,47 +24,60 @@
...
+                                                                        InstructorAttributes instructor) {
         boolean hasPublish = !session.isWaitingToOpen() && !session.isPublished();
         boolean hasUnpublish = !session.isWaitingToOpen() && session.isPublished();
         String disabledStr = "disabled=\"disabled\"";
-        String disableUnpublishSessionStr = instructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION) ? "" : disabledStr;
-        String disablePublishSessionStr = instructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION) ? "" : disabledStr;
+        String disableUnpublishSessionStr = instructor.isAllowedForPrivilege(Const.ParamsNames
+                                                                                  .INSTRUCTOR_PERMISSION_MODIFY_SESSION)
+                                          ? ""
+                                          : disabledStr;
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Doesn't look nice. May be something like this?
```
String disableUnpublishSessionStr = 
        instructor.isAllowedForPrivilege(Const.ParamsNames INSTRUCTOR_PERMISSION_MODIFY_SESSION)
            ? "" : disabledStr;

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsResponseRatePageAction.java [hunk 3]
```diff
@@ -8,28 +8,27 @@
...
+
+        InstructorFeedbackResultsResponseRatePageData data =
                 new InstructorFeedbackResultsResponseRatePageData(account);
         data.responseStatus = logic.getFeedbackSessionResponseStatus(feedbackSessionName, courseId);
         data.responseStatus.getStudentsWhoDidNotRespondToAnyQuestion();
-        
-        return createAjaxResult(Const.ViewURIs.INSTRUCTOR_FEEDBACK_RESULTS_BY_QUESTION, data);
+
+        return createAjaxResult(
+                Const.ViewURIs.INSTRUCTOR_FEEDBACK_RESULTS_BY_QUESTION, data);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsResponseRatePageAction.java]: why line break?

File: src/test/java/teammates/test/cases/ui/InstructorFeedbackResultsDownloadActionTest.java [hunk 4]
```diff
@@ -17,104 +17,103 @@
...
         InstructorFeedbackResultsDownloadAction action = getAction(paramsNormal);
         FileDownloadResult result = (FileDownloadResult) action.executeAndPostProcess();
-        
-        String expectedDestination = "filedownload?" +
-                "error=false" +
-                "&user=idOfInstructor1OfCourse1";
+
+        String expectedDestination = "filedownload?"
+                                   + "error=false"
+                                   + "&user=idOfInstructor1OfCourse1";
```
Comment (damithc) [inline src/test/java/teammates/test/cases/ui/InstructorFeedbackResultsDownloadActionTest.java]: No need for line breaks here?

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 5]
```diff
@@ -195,85 +199,88 @@ public void clickVisibilityOptionForResponseCommentAndSave(String idString, int
...
-        commentRow.findElement(By.tagName("form"))
-            .findElement(By.tagName("a")).click();
+        commentRow.findElement(By.tagName("form")).findElement(By.tagName("a")).click();
         ThreadHelper.waitFor(1500);
     }
-    
-    public void verifyCommentRowContent(String commentRowIdSuffix, String commentText, String giverName) {
+
+    public void verifyCommentRowContent(String commentRowIdSuffix,
+                                        String commentText, String giverName) {
```
Comment (damithc) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: The original is shorter than 110 chars

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 6]
```diff
@@ -195,85 +199,88 @@ public void clickVisibilityOptionForResponseCommentAndSave(String idString, int
...
         WebDriverWait wait = new WebDriverWait(browser.driver, 30);
         WebElement commentRow;
-        try{
-            commentRow = wait.until(ExpectedConditions.presenceOfElementLocated(By.id("responseCommentRow" + commentRowIdSuffix)));
-        } catch (TimeoutException e){
+        try {
+            commentRow = wait.until(ExpectedConditions.presenceOfElementLocated(By.id("responseCommentRow"
+                                                                                      + commentRowIdSuffix)));
```
Comment (damithc) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: This is OK too. (The break is at a higher-level than before although the indentation doesn't follow any rule; when you break a deeply-nested element, it becomes harder to read).
```
commentRow = wait.until(ExpectedConditions.presenceOfElementLocated(
                                                 By.id("responseCommentRow"+ commentRowIdSuffix)));
```

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 7]
```diff
@@ -281,18 +288,19 @@ public void clickViewPhotoLink(int panelBodyIndex, String urlRegex) throws Excep
...
-                browser.driver.findElements(By.cssSelector(".popover-content > img"))
-                              .get(browser.driver.findElements(By.cssSelector(".popover-content > img")).size() - 1)
-                              .getAttribute("src"));
-        
+
+        AssertHelper.assertContainsRegex(urlRegex,
+                                         browser.driver.findElements(By.cssSelector(".popover-content > img"))
+                                                       .get(browser.driver.findElements(By.cssSelector(".popover-content > img"))
+                                                                                          .size() - 1)
+                                                       .getAttribute("src"));
```
Comment (damithc) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: The new one is not more readable than the original. See if there is a better alternative.

**Comment by damithc** [comment]
added some comments.

**Comment by wkurniawan07** [comment]
Update (after a few days): Coverage is good, or accounted for whenever not good, except for these parts:
1. FeedbackSessionResponseStatus (37.8%)
Can be covered with unit test. The methods there are very generic (comparators and sorting) so I'm not sure if writing unit test for this is actually meaningful.
2. FeedbackSessionResultsBundle (35.4%)
Can be covered with unit tests. Has quite a large amount of comparators as well.
3. InstructorFeedbackResultsResponseRatePageAction (0%)
4. InstructorFeedbackResultsResponseRatePageData (0%)
I tried running full test without these two files, and still got dev green (on an unrelated now, it's another straight one). I guess I should just remove them?

**Comment by damithc** [comment]
Create separate issues for writing unit tests. No harm testing 'harmless-looking' comparators and sorting functions.

**Comment by wkurniawan07** [comment]
Before I merge this can I just check that it's safe to remove InstructorFeedbackResultsResponseRatePageAction and InstructorFeedbackResultsResponseRatePageData?

**Comment by kanghj** [comment]
Yes it is safe to remove them, they were formerly used to obtain the list of students on the bottom of the results page, but have been replaced in #2987

**Pull request merged**